Bluetooth: Controller: Only show one prompt for experimental configs

rugeGerritsen opened the pull request:
When browsing through the tree, you will now no longer see two entries
for each symbol that is experimental for BT_LL_SW_SPLIT

The prompt message for experimental messages have now been rephrased to
include (Split Link Layer)

See 
![image](https://user-images.githubusercontent.com/23234783/151816164-e5e9d362-95c3-4e9a-bfae-ad3dee70bf6c.png)

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -470,20 +470,17 @@ config BT_CTLR_CHAN_SEL_2
 
 config BT_CTLR_ADV_EXT
-	bool "LE Advertising Extensions" if !BT_LL_SW_SPLIT
+	bool "LE Advertising Extensions"
 	depends on BT_CTLR_ADV_EXT_SUPPORT
 	select BT_CTLR_CHAN_SEL_2 if BT_LL_SW_SPLIT && BT_BROADCASTER
 	select BT_CTLR_SCAN_REQ_NOTIFY if BT_LL_SW_SPLIT && BT_BROADCASTER
+	select EXPERIMENTAL if BT_LL_SW_SPLIT
```
I think this changes more than you expected. 

Didn't the first instance of `BT_CTLR_ADV_EXT` simply missing `[EXPERIMENTAL]` and `select EXPERIMENTAL`?

The option should be considered experimental, regardless of `BT_LL_SW_SPLIT` AFAIK. 

This comment applies to all the changes in this PR.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -470,20 +470,17 @@ config BT_CTLR_CHAN_SEL_2
 config BT_CTLR_ADV_EXT
-	bool "LE Advertising Extensions" if !BT_LL_SW_SPLIT
+	bool "LE Advertising Extensions"
 	depends on BT_CTLR_ADV_EXT_SUPPORT
 	select BT_CTLR_CHAN_SEL_2 if BT_LL_SW_SPLIT && BT_BROADCASTER
 	select BT_CTLR_SCAN_REQ_NOTIFY if BT_LL_SW_SPLIT && BT_BROADCASTER
+	select EXPERIMENTAL if BT_LL_SW_SPLIT
 	# Enable by default for BT_LL_SW_SPLIT when "LE Advertising Set Terminated event" is implemented
 	default y if BT_EXT_ADV && !BT_LL_SW_SPLIT
 	help
 	  Enable support for Bluetooth 5.0 LE Advertising Extensions in the
 	  Controller.
 
-config BT_CTLR_ADV_EXT
-	prompt "LE Advertising Extensions [EXPERIMENTAL]" if BT_LL_SW_SPLIT
-	select EXPERIMENTAL if BT_LL_SW_SPLIT
-
```
These changes are intentional:
- One declaration of `BT_CTLR_ADV_EXT` instead of two
- Move `select EXPERIMENTAL if BT_LL_SW_SPLIT` to the only declaration remaining. 

This makes it appear nicer than what is today (see attached screenshot)

The other alternative is to add `depends on BT_LL_SW_SPLIT` to the second declaration, but that would only solve the confusion partly. There would still be two entries turning up when searching.

alwa-nordic reviewed (dismissed):
Removing the label "[EXPERIMENTAL]" from the feature name is fine in my opinion. I agree it's better than showing it twice in search results.

alwa-nordic commented:
Adding label area DX to discuss:
 - The symbol appearing twice in search results, one with "[EXPERIMENTAL]" and one without is confusing to developers.
 - Arguably, it is the combination of two symbols that is "[EXPERIMENTAL]" here. Eg. `BT_LL_SW_SPLIT` (an implementation selection) and `BT_CTLR_ADV_EXT` (a feature). Is this well communicated to the developer by conditionally adding "[EXPERIMENTAL]" to the prompt of the feature symbol?

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -470,20 +470,17 @@ config BT_CTLR_CHAN_SEL_2
 
 config BT_CTLR_ADV_EXT
-	bool "LE Advertising Extensions" if !BT_LL_SW_SPLIT
+	bool "LE Advertising Extensions"
```
Wouldn't this still result in multiple entries of `BT_CTLR_ADV_EXT` when you search for it?

trond-snekvik replied:
Not sure about the search, but it will definitely break menuconfig. The rendering logic that produces the current duplicate entry thing in menuconfig is caused by menuconfig forcing menus with visible children into existence, even if they should be invisible. As the menu parent is always the last node before the `if`, the `[EXPERIMENTAL]` node is forced into existence, as it holds all the child nodes.

IMO, the right variant of this suggestion would be to keep the two entries, but remove the prompts for both, then create a single prompt entry below them that conditionally adds both prompts:

```
config BT_CTLR_ADV_EXT
    prompt "LE Advertising Extensions [EXPERIMENTAL]" if BT_LL_SW_SPLIT
    prompt "LE Advertising Extensions" 
```

This entry will only serve as a toggleable menu heading, really, but it should keep the logic of the other properties as-is.

cvinayak replied:
> Wouldn't this still result in multiple entries of `BT_CTLR_ADV_EXT` when you search for it?

There will be multiple entries when searched, but only one is enabled based on the `if` condition that is satisfied. This is similar to how all hidden Kconfig get listed when in search.

My suggestions do not have the original problem of duplicate entries when traversing using menuconfig/guiconfig.

> the right variant of this suggestion would be to keep the two entries, but remove the prompts for both, then create a single prompt entry below them that conditionally adds both prompts

This (multiple conditional prompt under an entry) is not permitted by Kconfig and fail to build with a warning.

rugeGerritsen replied:
So @cvinayak , @alwa-nordic, and @trond-snekvik . The remaining question is: Are we ok with having two entries? It looks a bit messy IMO. But the proposed solution is better than the existing one

cvinayak replied:
It is ok to have see two prompt messages, and is consistent with the Link Layer selection wherein only one is selectable based on the Link Layer enabled.

alwa-nordic commented:
@cvinayak You NAK-ed this PR, but it's unclear to both the author and me what your objections are and how we should proceed. Could you have a look at this again?

cvinayak commented:
> @cvinayak You NAK-ed this PR, but it's unclear to both the author and me what your objections are, and so we don't know how to proceed. Could you have a look at this again?

@alwa-nordic  The title of the PR and the description do not describe the entirety of the problem. One, duplicate prompt when traversing through the Kconfig menu items, and second, display of duplicate prompt when searching. My suggestion in this PR addresses the former. The later, to have duplicate prompts when searching is acceptable and is inline with other hidden Kconfig that get listed.

tejlmand reviewed (changes requested):
Having both prompts appearing in the search results is unfortunate, but the side-effects of this proposal are worse, when considering UI interface of items in menuconfig and doc generation.

Unfortunately I don't have a good solution to the search issue.

rugeGerritsen commented:
> 

@cvinayak , thanks, that works out nicely. I must have missed something when I tried it the first time

tejlmand reviewed (commented):
small suggestion regarding the duplicate search experimental confusion.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -525,9 +528,12 @@ config BT_CTLR_ADV_PERIODIC
 
 config BT_CTLR_ADV_PERIODIC
-	bool "LE Periodic Advertising in Advertising State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+	prompt "LE Periodic Advertising in Advertising State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
```
Good point. I will let @cvinayak decide on that

cvinayak replied:
We can have it postfix, `LE Periodic Advertising in Advertising State (Split Link Layer) [EXPERIMENTAL] `

cvinayak replied:
One other preference, to keep the `prompt` separate and retain to keep the dependencies/select along with the Kconfig that adds `bool`.

rugeGerritsen replied:
Where should the help text be placed. AFAIK it needs to be added to a prompt

cvinayak replied:
Help text can remain with `bool`, Kconfig options append each properties irrespective of which Kconfig file appends it, i.e. same Kconfig with different properties placed in application, board, architecture, SoC Kconfig file get appended in the end.

tejlmand reviewed (changes requested):
could you please update the pr description and commit message as to what you want to achieve with this PR.

As I see it, there are no changes to visibility compared to existing code, but creating a symbol independently as this PR now does actually means that `depends on` doesn't have the same effect any more.

This is not a problem in this particular case, as the existing code doesn't have extra `depends on`, but I start to wonder, should the existing:
```
config BT_CTLR_ADV_EXT
	prompt "LE Advertising Extensions [EXPERIMENTAL]" if BT_LL_SW_SPLIT
	select EXPERIMENTAL if BT_LL_SW_SPLIT
	depends on BT_LL_SW_SPLIT
```

tejlmand commented:
Seem a comment of mine got lost.

Regarding the visibility, this is what i see on `zephyr/main`:
![image](https://user-images.githubusercontent.com/9802583/153389821-2f2c98da-7089-4f74-aa1f-e9a1a8e89b6b.png)

and with this PR:
![image](https://user-images.githubusercontent.com/9802583/153389990-6e4f2c66-ff03-40b6-9e3a-bcb3d175b148.png)

tejlmand commented:
> but creating a symbol independently as this PR now does actually means that depends on doesn't have the same effect any more.

To exemplify, try the following code:
```
config BAR
        bool

config FOO
        bool
        prompt "Test FOO" if BAR
        depends on BAR

config BAZ
        bool
        default y
        select FOO
```

vs.
```
config BAR
        bool

config FOO
        bool

config FOO
        prompt "Test FOO" if BAR
        depends on BAR

config BAZ
        bool
        default y
        select FOO
```

first version will fail, because `BAZ` tries to select `FOO`, but `FOO` depends on `BAR` which is `n`.

Second variant will pass, because now `FOO` can be selected because the `config FOO \n bool` doesn't have any dependencies, only the prompt version has.

So be very careful when starting to split settings and their prompt, due to dependencies.

tejlmand reviewed (changes requested):
you are changing the dependencies of those settings without explaining why that is needed.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -536,18 +542,21 @@ config BT_CTLR_ADV_PERIODIC_ADI_SUPPORT
 
 config BT_CTLR_SYNC_PERIODIC
-	bool "LE Periodic Advertising in Synchronization State" if !BT_LL_SW_SPLIT
+	prompt "LE Periodic Advertising in Synchronization State" if !BT_LL_SW_SPLIT
```
@tejlmand , I think you tried this with one of the older commits pushed to this PR. I'm not able to reproduce what you got. This is what I did:
```
cd <workspace>/nrf 
git checkout 7371dec6308cea591bc1b450f1710f0c73b60b87
west config manifest.path nrf
west update
cd <workspace>/zephyr
git cherry-pick a82f92995207e5101eacb365710f963f1d700c90
```
With BT_LL_SW_SPLIT:
```
$ west build -b nrf52840dk_nrf52840 samples/bluetooth/periodic_adv/ -- -DCONFIG_BT_LL_SW_SPLIT=y -DCONFIG_BT_CTLR_SYNC_PERIODIC=y
-- west build: generating a build system
Including boilerplate (Zephyr base): C:/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: C:/ncs/zephyr/samples/bluetooth/periodic_adv
-- Zephyr version: 2.7.99 (C:/ncs/zephyr), build: v2.7.99-ncs1-rc1-69-gc9a09ec4fc13
-- Found Python3: C:/Python36/python3.exe (found suitable exact version "3.6.8") found components: Interpreter
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- Cache files will be written to: C:\Users\ruge\AppData\Local/.cache/zephyr
-- Found dtc: C:/ProgramData/chocolatey/bin/dtc.exe (found suitable version "1.4.7", minimum required is "1.4.6")
-- Found toolchain: gnuarmemb (C:/Program Files (x86)/GNU Tools ARM Embedded/8 2019-q3-update)
-- Found BOARD.dts: C:/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Generated zephyr.dts: C:/ncs/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: C:/ncs/zephyr/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: C:/ncs/zephyr/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: C:/ncs/zephyr/build/zephyr/dts.cmake

warning: Experimental symbol BT_LL_SW_SPLIT is enabled.


warning: Experimental symbol BT_CTLR_ADV_EXT is enabled.


warning: Experimental symbol BT_CTLR_ADV_PERIODIC is enabled.


warning: Experimental symbol BT_CTLR_SYNC_PERIODIC is enabled.

Parsing C:/ncs/zephyr/Kconfig
Loaded configuration 'C:/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration 'C:/ncs/zephyr/samples/bluetooth/periodic_adv/prj.conf'
Merged configuration 'C:/ncs/zephyr/samples/bluetooth/periodic_adv/boards/nrf52840dk_nrf52840.conf'
Merged configuration 'C:/ncs/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to 'C:/ncs/zephyr/build/zephyr/.config'
Kconfig header saved to 'C:/ncs/zephyr/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is GNU 8.3.1
-- The ASM compiler identification is GNU
-- Found assembler: C:/Program Files (x86)/GNU Tools ARM Embedded/8 2019-q3-update/bin/arm-none-eabi-gcc.exe
CMake Warning at C:/ncs/zephyr/subsys/bluetooth/controller/CMakeLists.txt:29 (message):


  CONFIG_BT_CTLR_ADV_EXT=y, Advertising Extensions Feature in Zephyr
  Bluetooth Controller is EXPERIMENTAL.


-- Configuring done
-- Generating done
-- Build files have been written to: C:/ncs/zephyr/build
-- west build: building application
[1/225] Generating misc/generated/syscalls.json, misc/generated/struct_tags.json
[2/225] Generating include/generated/kobj-types-enum.h, include/generated/otype-to-str.h, include/generated/otype-to-size.h
```
With BT_LL_SOFTDEVICE:
```
$ west build -b nrf52840dk_nrf52840 samples/bluetooth/periodic_adv/ -- -DCONFIG_BT_LL_SOFTDEVICE=y -DCONFIG_BT_CTLR_SYNC_PERIODIC=y
-- west build: generating a build system
Including boilerplate (Zephyr base): C:/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: C:/ncs/zephyr/samples/bluetooth/periodic_adv
-- Zephyr version: 2.7.99 (C:/ncs/zephyr), build: v2.7.99-ncs1-rc1-69-gc9a09ec4fc13
-- Found Python3: C:/Python36/python3.exe (found suitable exact version "3.6.8") found components: Interpreter
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- Cache files will be written to: C:\Users\ruge\AppData\Local/.cache/zephyr
-- Found dtc: C:/ProgramData/chocolatey/bin/dtc.exe (found suitable version "1.4.7", minimum required is "1.4.6")
-- Found toolchain: gnuarmemb (C:/Program Files (x86)/GNU Tools ARM Embedded/8 2019-q3-update)
-- Found BOARD.dts: C:/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Generated zephyr.dts: C:/ncs/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: C:/ncs/zephyr/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: C:/ncs/zephyr/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: C:/ncs/zephyr/build/zephyr/dts.cmake
Parsing C:/ncs/zephyr/Kconfig
Loaded configuration 'C:/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration 'C:/ncs/zephyr/samples/bluetooth/periodic_adv/prj.conf'
Merged configuration 'C:/ncs/zephyr/samples/bluetooth/periodic_adv/boards/nrf52840dk_nrf52840.conf'
Merged configuration 'C:/ncs/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to 'C:/ncs/zephyr/build/zephyr/.config'
Kconfig header saved to 'C:/ncs/zephyr/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is GNU 8.3.1
-- The ASM compiler identification is GNU
-- Found assembler: C:/Program Files (x86)/GNU Tools ARM Embedded/8 2019-q3-update/bin/arm-none-eabi-gcc.exe
-- Configuring done
-- Generating done
-- Build files have been written to: C:/ncs/zephyr/build
-- west build: building application
[1/202] Generating misc/generated/syscalls.json, misc/generated/struct_tags.json
```

tejlmand replied:
@rugeGerritsen you're proving my point.

You are **supposed** to get an error.
I wrote:
> then I can suddenly set this variable.

User's are **not** supposed to be able to enable that setting, but with your changes they can.
Try the same on `ncs/main` and you'll see that very same build command will report an error to the user.

rugeGerritsen replied:
@tejlmand , sorry, I don't understand what you are trying to point out. Users **are** supposed to be able to set `BT_CTLR_SYNC_PERIODIC`. 

The logs I posted above are from ncs/main (otherwise I would not be able to use CONFIG_BT_LL_SOFTDEVICE)

tejlmand replied:
No, as soon as you do this:
```
cd <workspace>/zephyr
git cherry-pick a82f92995207e5101eacb365710f963f1d700c90
```

then you have your changes on top

rugeGerritsen replied:
@tejlmand , sorry for my lack of understanding. Can you briefly try to point out:
1. What are you trying to achieve. 
2. What is the expected output
3. What do you actually see.

I don't understand what you are trying to say with
> No, as soon as you do this:
> cd <workspace>/zephyr
> git cherry-pick a82f92995207e5101eacb365710f963f1d700c90
> then you have your changes on top

tejlmand replied:
Hint, this indicates you are building with your cherry-pick commit:
> -- Zephyr version: 2.7.99 (C:/ncs/zephyr), build: v2.7.99-ncs1-rc1-69-gc9a09ec4fc13

Cause `c9a09ec4fc13` is not a valid SHA here https://github.com/nrfconnect/sdk-zephyr, hence it must be a SHA only available at your local workspace.

tejlmand replied:
> What are you trying to achieve.

I'm concerned that you accidentally are changing the dependencies of those symbols when tryin to solve an issue that is not there in the first place. See this message: https://github.com/zephyrproject-rtos/zephyr/pull/42319#issuecomment-1034759189

So i'm trying to avoid a PR going in, which suddenly allows users to enable settings they are not supposed to enable.

> What is the expected output

According to you commit message, this is the expected output:
>  Bluetooth: Controller: Only show one prompt for experimental configs
>
> When browsing through the tree, you will now no longer see two entries
> for each symbol that is experimental for BT_LL_SW_SPLIT
>
> The prompt message for experimental messages have now been rephrased to
> include (Split Link Layer)

and therefore I don't assume kconfig dependencies should change.

> What do you actually see.

Changed kconfig dependencies.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -58,6 +58,33 @@ config BT_LLL_VENDOR_OPENISA
 
+# First redefine the experimental symbols. 
+# We define them here due to avoid declaring them right below the non-experimental
+# definitions. In that case, both would show up, which is an issue in the Kconfig library.
```
@tejlmand , as discussed it would be nice if we could add a reference to the KConfig issue here.

tejlmand replied:
yep, still trying to reproduce using a small snippet to fully understand the nature of the problem.

tejlmand replied:
See issue / explanation of this behavior here https://github.com/ulfalizer/Kconfiglib/issues/117

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -58,6 +58,33 @@ config BT_LLL_VENDOR_OPENISA
+# When the feature is no longer experimental, simply remove the line here
+# and update the original implementations.
+
+config BT_CTLR_ADV_EXT
+	prompt "LE Advertising Extensions (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_ADV_PERIODIC
+	bool "LE Periodic Advertising in Advertising State[EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_SYNC_PERIODIC
+	bool "LE Periodic Advertising in Synchronization State (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_ADV_ISO
+	bool "LE Broadcast Isochronous Channel advertising (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_SYNC_ISO
+	bool "LE Broadcast Isochronous Channel advertising sync (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_CENTRAL_ISO
+	bool "LE Connected Isochronous Stream Central (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
+config BT_CTLR_PERIPHERAL_ISO
+	bool "LE Connected Isochronous Stream Peripheral (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+
```
```suggestion
config BT_CTLR_ADV_EXT
	prompt "LE Advertising Extensions (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_ADV_PERIODIC
	prompt "LE Periodic Advertising in Advertising State[EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_SYNC_PERIODIC
	prompt "LE Periodic Advertising in Synchronization State (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_ADV_ISO
	prompt "LE Broadcast Isochronous Channel advertising (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_SYNC_ISO
	prompt "LE Broadcast Isochronous Channel advertising sync (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_CENTRAL_ISO
	prompt "LE Connected Isochronous Stream Central (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

config BT_CTLR_PERIPHERAL_ISO
	prompt "LE Connected Isochronous Stream Peripheral (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT

```

cvinayak reviewed (changes requested):
~With the latest changes, none of these Kconfigs are selectable in menuconfig with upstream repo.~ Now the duplicate entry are far down the list in the upstream repo.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -62,6 +62,40 @@ config BT_LLL_VENDOR_OPENISA
 	  Use OpenISA Lower Link Layer implementation.
 
+# First redefine the experimental symbols.
+# We define them here due to avoid declaring them right below the non-experimental
```
Good point. Should we rather decline this PR and resolve the Kconfig issue then?

tejlmand replied:
> Should we rather decline this PR and resolve the Kconfig issue then?

at this point, i'm not confident that the is a easy solution to the Kconfiglib issue.
Cause trying to solve that issue might easily break this functionality:
```
# Show invisible symbols if they have visible children. This 
# can happen for an m/y-valued symbol with an optional prompt 
# ('prompt "foo" is COND') that is currently disabled. Note 
# that it applies to both 'config' and 'menuconfig' symbols. 
```

in which case we might end-up changing one kind of misbehaving with another kind.

However, I also posted a workaround here: https://github.com/ulfalizer/Kconfiglib/issues/117
which you can try out, and see if it creates the desired result.

rugeGerritsen replied:
I tried it out, it didn't work (pushed the changes to demonstrate it)

rugeGerritsen committed:
Bluetooth: Controller: Only show one prompt for experimental configs

Uses the proposed workaround in
https://github.com/ulfalizer/Kconfiglib/issues/117

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
Bluetooth: Controller: Enhance experimental SW_SPLIT Kconfigs names

This makes the it more intuitive when searching.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

tejlmand commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -682,15 +712,21 @@ config BT_CTLR_SET_HOST_FEATURE
 config BT_CTLR_CENTRAL_ISO
 	bool "LE Connected Isochronous Stream Central" if !BT_LL_SW_SPLIT
 	depends on BT_CTLR_CENTRAL_ISO_SUPPORT && BT_CENTRAL
+	select BT_CTLR_SET_HOST_FEATURE
```
curious. what does this change has to do with ?
> Only show one prompt for experimental configs

tejlmand reviewed (changes requested):
could you please checkup whether the workaround is needed in all cases ?

From what I can tell, only a few places suffers from the problem.
Please read https://github.com/ulfalizer/Kconfiglib/issues/117 carefully to understand under which circumstances this problem manifests itself.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -682,15 +712,21 @@ config BT_CTLR_SET_HOST_FEATURE
+	help
+	  Unused and hidden Kconfig symbol to allow prompt adjustment.
+	  Work around for https://github.com/ulfalizer/Kconfiglib/issues/117
+
```
Removed

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -609,9 +633,15 @@ config BT_CTLR_SYNC_ISO
+	help
+	  Unused and hidden Kconfig symbol to allow prompt adjustment.
+	  Work around for https://github.com/ulfalizer/Kconfiglib/issues/117
+
```
Removed

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -596,9 +614,15 @@ config BT_CTLR_ADV_ISO
+	help
+	  Unused and hidden Kconfig symbol to allow prompt adjustment.
+	  Work around for https://github.com/ulfalizer/Kconfiglib/issues/117
+
```
Removed

tejlmand commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -525,7 +531,7 @@ config BT_CTLR_ADV_PERIODIC
 
 config BT_CTLR_ADV_PERIODIC
-	bool "LE Periodic Advertising in Advertising State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+	bool "LE Periodic Advertising in Advertising State (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
```
comment not directly related to this line, but cannot comment below.

Why was this removed in last update ?
https://github.com/zephyrproject-rtos/zephyr/blob/65a659e7f7e2ad76a2406f8d6db57bca99ee25b3/subsys/bluetooth/controller/Kconfig#L537-L538

Only the help text was supposed to be remove if you read this comment https://github.com/zephyrproject-rtos/zephyr/pull/42319#discussion_r811024770.

This code part still suffers the double prompt issue, as there are two `config BT_CTLR_ADV_PERIODIC` after one another, and then a `depends on BT_CTLR_ADV_PERIODIC` here:
https://github.com/zephyrproject-rtos/zephyr/blob/e4e8b78f93a7871c4494bf6b2e5596e97fcf7c12/subsys/bluetooth/controller/Kconfig#L524-L539

Especially pay attention to this line:
https://github.com/zephyrproject-rtos/zephyr/blob/e4e8b78f93a7871c4494bf6b2e5596e97fcf7c12/subsys/bluetooth/controller/Kconfig#L539

Show you still need
```
config PROMPT_ADJUSTMENT_WORKAROUND
	bool
```
just no help text.

tejlmand commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -545,7 +551,7 @@ config BT_CTLR_SYNC_PERIODIC
 
 config BT_CTLR_SYNC_PERIODIC
-	bool "LE Periodic Advertising in Synchronization State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
+	bool "LE Periodic Advertising in Synchronization State (Split Link Layer) [EXPERIMENTAL]" if BT_LL_SW_SPLIT
```
same here. not related to this line, but below.

You still need:
https://github.com/zephyrproject-rtos/zephyr/blob/65a659e7f7e2ad76a2406f8d6db57bca99ee25b3/subsys/bluetooth/controller/Kconfig#L563-L564

just without the help text.

Notice the two config BT_CTLR_SYNC_PERIODIC after one another, and then a depends on BT_CTLR_SYNC_PERIODIC here:
https://github.com/zephyrproject-rtos/zephyr/blob/e4e8b78f93a7871c4494bf6b2e5596e97fcf7c12/subsys/bluetooth/controller/Kconfig#L544-L559

tejlmand reviewed (changes requested):
A little too much was remove in last push.

Please notice I only asked for removal of the help text those two places:
https://github.com/zephyrproject-rtos/zephyr/pull/42319#discussion_r811024770
https://github.com/zephyrproject-rtos/zephyr/pull/42319#discussion_r811025231

The
```
config PROMPT_ADJUSTMENT_WORKAROUND
	bool
```
must still be there.

rugeGerritsen commented:
@tejlmand , thanks for the thorough review

tejlmand reviewed (approved):
lgtm - Thanks for the effort and patience.

This sure was a tricky issue, but the end-result looks good.

cvinayak reviewed (approved):
Approved the prompt text changes. Not reviewed/tested for the prompt workaround.

rugeGerritsen commented:
I believe this PR can be merged now

rugeGerritsen commented:
@carlescufi , can you merge this PR?

nashif commented:
this is marked as DNM, it will not be merged with this label.
